iree/compiler/Dialect/Flow/IR/FlowOpFolders.cpp

```diff
@@ -253,6 +219,157 @@ void VariableStoreIndirectOp::getCanonicalizationPatterns(
   results.insert<PropagateVariableStoreAddress>(context);
 }
 
+//===----------------------------------------------------------------------===//
+// Dispatch ops
+//===----------------------------------------------------------------------===//
+
+namespace {
+
+struct DceDispatchRegion : public OpRewritePattern<DispatchRegionOp> {
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(DispatchRegionOp op,
+                                PatternRewriter &rewriter) const override {
+    if (op.body().empty()) return failure();
+    ClosureOpDce dce(op, op.body().front(), /*variadicOffset=*/1);
```

Comment:
It unfortunately has some issues here that prevent us from using it for the new dispatch op (it requires both the inside and outside have 1:1 args/results, which we don't have) - haven't looked into seeing if there's parts of it we can reuse or not yet, but it'd be nice. Particularly for when multiple dynamic shape dimensions all have the same value (like when shapes get transposed/etc) to prevent us from needing to pass so many of them.